caa: Add option to skip Peer pod VM delete

stevenhorsman opened the pull request:
During development, several of us have hit times where being able to access a failing peer pod VM has been crucial to debugging issues. The general approach to this is to add a sleep before, or comment out `DeleteInstance`, but this involved rebuilding the CAA image, so instead we can provide an option to set `PEERPODS_DEVELOPER_MODE=true` in peer-pods-cm which will skip this peer pod vm delete.

stevenhorsman committed:
caa: Add option to skip Peer pod VM delete

During development, several of us have hit times where being able to
access a failing peer pod VM has been crucial to debugging issues.
The general approach to this is to add a sleep before, or comment out
`DeleteInstance`, but this involved rebuilding the CAA image, so instead
we can provide an option to set `PEERPODS_DEVELOPER_MODE=true`
in peer-pods-cm which will skip this peer pod vm delete.

Signed-off-by: stevenhorsman <steven@uk.ibm.com>

bpradipt commented:
@stevenhorsman wondering if for developer mode, the per node limit be set to 1 to avoid K8s recreating new instances?

bpradipt commented:
@stevenhorsman any plans to proceed with this ?

stevenhorsman commented:
> @stevenhorsman any plans to proceed with this ?

Yes, but it's not high priority and the extra scope creep meant it needed more attention than I have time to dedicate at the moment.

beraldoleal reviewed (approved):
Useful and LGTM. Thanks @stevenhorsman

stevenhorsman commented:
I've rebase and re-tested this manually and it seems like the delete is correctly blocked when in dev mode:
```
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kubectl get cm peer-pods-cm -n confidential-containers-system -o json | grep PEERPODS_DEVELOPER_MODE
        "PEERPODS_DEVELOPER_MODE": "true",
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kubectl get pod
NAME    READY   STATUS    RESTARTS   AGE
nginx   1/1     Running   0          8m40s
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kcli list vms
+----------------------+--------+-----------------+------------+-----------+---------+
|         Name         | Status |        Ip       |   Source   |    Plan   | Profile |
+----------------------+--------+-----------------+------------+-----------+---------+
| peer-pods-ctlplane-0 |   up   | 192.168.122.155 | ubuntu2204 | peer-pods |  kvirt  |
|  peer-pods-worker-0  |   up   |  192.168.122.50 | ubuntu2204 | peer-pods |  kvirt  |
| podvm-nginx-3f105dc8 |   up   |  192.168.122.40 |            |           |         |
+----------------------+--------+-----------------+------------+-----------+---------+
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kubectl delete -f ibmcloud/demo/nginx.yaml
pod "nginx" deleted
configmap "nginx-config" deleted
secret "nginx-secret" deleted
service "nginx-service" deleted
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kubectl get pod
No resources found in default namespace.
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kcli list vms
+----------------------+--------+-----------------+------------+-----------+---------+
|         Name         | Status |        Ip       |   Source   |    Plan   | Profile |
+----------------------+--------+-----------------+------------+-----------+---------+
| peer-pods-ctlplane-0 |   up   | 192.168.122.155 | ubuntu2204 | peer-pods |  kvirt  |
|  peer-pods-worker-0  |   up   |  192.168.122.50 | ubuntu2204 | peer-pods |  kvirt  |
| podvm-nginx-3f105dc8 |   up   |  192.168.122.40 |            |           |         |
+----------------------+--------+-----------------+------------+-----------+---------+
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kubectl apply -f ibmcloud/demo/nginx.yaml
pod/nginx created
configmap/nginx-config created
secret/nginx-secret created
service/nginx-service created
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kubectl get pod
NAME    READY   STATUS              RESTARTS   AGE
nginx   0/1     ContainerCreating   0          4s
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kcli list vms
+----------------------+--------+-----------------+------------+-----------+---------+
|         Name         | Status |        Ip       |   Source   |    Plan   | Profile |
+----------------------+--------+-----------------+------------+-----------+---------+
| peer-pods-ctlplane-0 |   up   | 192.168.122.155 | ubuntu2204 | peer-pods |  kvirt  |
|  peer-pods-worker-0  |   up   |  192.168.122.50 | ubuntu2204 | peer-pods |  kvirt  |
| podvm-nginx-3f105dc8 |   up   |  192.168.122.40 |            |           |         |
| podvm-nginx-4619a474 |   up   |                 |            |           |         |
+----------------------+--------+-----------------+------------+-----------+---------+
```

For completeness I turned `PEERPODS_DEVELOPER_MODE` back off to check that the basic behaviour works:
```
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kubectl get cm peer-pods-cm -n confidential-containers-system -o json | grep PEERPODS_DEVELOPER_MODE
        "PEERPODS_DEVELOPER_MODE": "false",
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kubectl get pods
NAME    READY   STATUS    RESTARTS   AGE
nginx   1/1     Running   1          14m
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kcli list vm
+----------------------+--------+-----------------+------------+-----------+---------+
|         Name         | Status |        Ip       |   Source   |    Plan   | Profile |
+----------------------+--------+-----------------+------------+-----------+---------+
| peer-pods-ctlplane-0 |   up   | 192.168.122.155 | ubuntu2204 | peer-pods |  kvirt  |
|  peer-pods-worker-0  |   up   |  192.168.122.50 | ubuntu2204 | peer-pods |  kvirt  |
| podvm-nginx-ef1a00cb |   up   | 192.168.122.196 |            |           |         |
+----------------------+--------+-----------------+------------+-----------+---------+
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kubectl delete -f ibmcloud/demo/nginx.yaml
pod "nginx" deleted
configmap "nginx-config" deleted
secret "nginx-secret" deleted
service "nginx-service" deleted
root@sh-caa-dev-mode1:~/go/src/github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor# kcli list vm
+----------------------+--------+-----------------+------------+-----------+---------+
|         Name         | Status |        Ip       |   Source   |    Plan   | Profile |
+----------------------+--------+-----------------+------------+-----------+---------+
| peer-pods-ctlplane-0 |   up   | 192.168.122.155 | ubuntu2204 | peer-pods |  kvirt  |
|  peer-pods-worker-0  |   up   |  192.168.122.50 | ubuntu2204 | peer-pods |  kvirt  |
+----------------------+--------+-----------------+------------+-----------+---------+
```

stevenhorsman committed:
caa: If in developer mode set default peerpods-limit-per-node 1

If we are in developer mode then set the default
peerpods-limit-per-node to be 1 to avoid Kubernetes
creating multiple instances of a pod which has issues.

Signed-off-by: stevenhorsman <steven@uk.ibm.com>

stevenhorsman commented:
> @stevenhorsman wondering if for developer mode, the per node limit be set to 1 to avoid K8s recreating new instances?

@bpradipt - I have added an extra commit for this, but don't have time to test it at the moment, so let me know if you are okay for this to be merged as-is, want the commit removed from the PR, or want to wait. Thanks

ldoktor commented:
@stevenhorsman is this supposed to work for any provider? I gave it a try with aws and it removed the VMs. (note I deployed it with the default caa-daemonset and then modified it to use the one built from this PR sources, not sure if that is the problem)

ldoktor commented:
> > @stevenhorsman is this supposed to work for any provider? I gave it a try with aws and it removed the VMs. (note I deployed it with the default caa-daemonset and then modified it to use the one built from this PR sources, not sure if that is the problem)
> 
> Yeah, it should work with any provider. I don't have access to AWS though, so can't debug this

I gave it a few attempts and even got a VM left-behind running by apply followed immediately by delete. But when it successfully boots or I get container creating problem (eg. by running non-existing command) it shuts the VM down.
[cloud-api-adaptor-daemonset.log](https://github.com/user-attachments/files/19340152/cloud-api-adaptor-daemonset.log)

Therefore it does work when one wants to debug issue where the VM won't boot, but when it connects to it, it will shut it down anyway despite the message.

stevenhorsman commented:
> Therefore it does work for cases where the VM won't boot at all, but whenever caa connects to it, it will shut it down no matter what the "Running in developer mode, so leaving instance i-0417acf40affe2bf6 running" message says on AWS.

Interesting - with my tested example on libvirt I used a working pod, so it went into running state, and then wasn't cleaned up, which sounds like it isn't working. I can try making a bad image that doesn't connect though as that's the main use-case here.

PS: Sorry about pushing the commit to the wrong branch!

ldoktor commented on src/cloud-api-adaptor/entrypoint.sh:
```diff
@@ -33,7 +33,12 @@ optionals+=""
+if [[ "${PEERPODS_DEVELOPER_MODE}" == "true" ]]; then
+    optionals+="-developer-mode "
+    if [[ -z "${PEERPODS_LIMIT_PER_NODE}" ]]; then
+        optionals+="-peerpods-limit-per-node 1 "
```
Well this is tricky as usually this limit is set. So on top of modifying PEERPODS_DEVELOPER_MODE one would have to remove the PEERPODS_LIMIT_PER_NODE in order for this to work. And if one has to do that, they can as well simply set it to the expected number, what do you think?

On the other hand always changing it might result in even worse problems so I think adding a comment to PEERPODS_DEVELOPER_MODE about a good idea of setting PEERPODS_LIMIT_PER_NODE to 1 would have similar value and would be more explicit. What do you think?

ldoktor replied:
Btw does this work? Because on AWS when I set this to `1` and the VM fails to boot, it shuts the (booted but unable to access) VM down and replaces it with a new one.

```
2025/03/19 12:40:37 [adaptor/proxy] Retrying failed agent proxy connection: dial tcp 3.238.157.120:15150: connect: connection refused
2025/03/19 12:40:42 [adaptor/proxy] Retrying failed agent proxy connection: dial tcp 3.238.157.120:15150: connect: connection refused
2025/03/19 12:40:47 [adaptor/proxy] Retrying failed agent proxy connection: dial tcp 3.238.157.120:15150: connect: connection refused
2025/03/19 12:40:48 [adaptor/cloud] Error: start instance interrupted (context canceled). Cleaning up...
2025/03/19 12:40:48 [adaptor/proxy] shutting down socket forwarder
2025/03/19 12:40:48 [adaptor/cloud] error starting instance: context canceled
2025/03/19 12:40:50 [podnetwork] routes on netns /var/run/netns/cni-93d5a334-0c9f-6da1-47d1-e67767c38208
2025/03/19 12:40:50 [podnetwork]     0.0.0.0/0 via 10.244.1.1 dev eth0
2025/03/19 12:40:50 [podnetwork]     10.244.0.0/16 via 10.244.1.1 dev eth0
2025/03/19 12:40:50 [podnetwork]     10.244.1.0/24  dev eth0
2025/03/19 12:40:50 [adaptor/cloud] stored /run/peerpod/pods/e1d58f0a4bba892eb74cba3f8cfa5cc33d7ab39a4273024e15c52c15f57a8a66/daemon.json
2025/03/19 12:40:50 [adaptor/cloud] error reading image pull secrets: serviceaccounts "default" is forbidden: User "system:serviceaccount:confidential-containers-system:cloud-api-adaptor" cannot get resource "serviceaccounts" in API group "" in the namespace "confidential-containers-system"
2025/03/19 12:40:50 [adaptor/cloud] initdata in Pod annotation: 
2025/03/19 12:40:50 [adaptor/cloud] initdata in pod annotation is empty, use global initdata: 
2025/03/19 12:40:50 [adaptor/cloud] create a sandbox e1d58f0a4bba892eb74cba3f8cfa5cc33d7ab39a4273024e15c52c15f57a8a66 for pod http-server in namespace confidential-containers-system (netns: /var/run/netns/cni-93d5a334-0c9f-6da1-47d1-e67767c38208)
2025/03/19 12:40:50 [adaptor/cloud] Using default instance type ("t3.small")
2025/03/19 12:40:50 [adaptor/cloud/aws] Creating instance podvm-http-server-e1d58f0a for sandbox e1d58f0a4bba892eb74cba3f8cfa5cc33d7ab39a4273024e15c52c15f57a8a66
2025/03/19 12:40:51 [adaptor/cloud/aws] Created instance podvm-http-server-e1d58f0a (i-00b3baaf0f84eff2d) for sandbox e1d58f0a4bba892eb74cba3f8cfa5cc33d7ab39a4273024e15c52c15f57a8a66
```

So it looks like AWS provider tries really hard not to leaving any VMs behind... The only way to leave them I found was to have not-yet-booted/unbootable VM when we delete the pod. Then it stays forever, but whenever it connects to it or when it decides it's unbootable, it just shuts it down.

Still it's a useful feature as previously we were not able to leave the VM running (unless one forcefully destroyed the operator :D)

bpradipt replied:
The default peer-pod-limit is 10.

@ldoktor when you say "On the other hand always changing it might result in even worse problems", what exactly are you referring to..

bpradipt replied:
For PEERPODS_LIMIT_PER_NODE to work, you will need to use webhook as well. 
I'm ok to document it. 
@ldoktor when you tested this on AWS with a regular AMI (not having kata), did K8s tried to spin up new pods on error ?

bpradipt replied:
PEERPODS_LIMIT_PER_NODE has no effect if you have not installed the webhook.

ldoktor replied:
Hmm, the `restartPolicy: Never ` is actually not reflected on podvm boot failure on AWS. I have the webhook deployed (`CLOUD_PROVIDER=aws make deploy`) and the PEERPODS_LIMIT_PER_NODE also won't influence the recreation (again only tested on AWS).

chathuryaadapa commented:
1. Environment Setup Completed
   - Following the repository's libvirt documentation (libvirt/README.md, config_libvirt.sh, kcli_cluster.sh, and install_caa.sh):
   - Prerequisites & Libvirt: Installed Go, Helm, libvirt, QEMU-KVM, andkcli.
   - Kubernetes Cluster: Created a 2-node cluster withkcli:
```
peer-pods-ctlplane-0
peer-pods-worker-0
```
   - PodVM Image: Extracted and registeredpodvm-base.qcow2into the libvirt storage pool.
   - Cloud-API-Adaptor Build: Built a dev container image (cloud-api-adaptor:local-test) containing the exact code from this PR branch and loaded it directly into containerd on the worker node.
   - CAA & Kata Deploy: Deployed the Confidential Containers Helm chart (kata-remoteruntimeClass,kata-deploy,peerpodctrl,cloud-api-adaptor).


2. Test Execution & Evidence

**Test 1: Normal Mode (Regression Test)**
   - Configuration: Default (PEERPODS_DEVELOPER_MODEnot set).
     Verification:
      - CAA started with default limit `PeerPodsLimitPerNode: 10.`
      - Node extended resource reported `[kata.peerpods.io/vm](http://kata.peerpods.io/vm): 10` capacity.
      - Created peer `podtest1-busybox`, Libvirt VM `podvm-test1-busybox-8802d4c1` was instantiated.
      - Deleted the pod , CAA deleted the Libvirt VM and released the PeerPod extended resource (0allocated).

**Test 2: Developer Mode + Default Limit**
     - Configuration: `PEERPODS_DEVELOPER_MODE: "true"` in `peer-pods-cm`
     Verification:
```
kubectl logs -n confidential-containers-system -l app=cloud-api-adaptor  

 + exec cloud-api-adaptor libvirt -data-dir /opt/data-dir -developer-mode -**peerpods-limit-per-node 1**
2026/08/27 13:43:58 [adaptor/cloud] Cloud provider external plugin loading is disabled, skipping plugin loading
2026/08/27 13:43:58 [adaptor/cloud/libvirt] libvirt config: &libvirt.Config{URI:"qemu+ssh://root@192.168.122.1/system?no_verify=1", ...}
cloud-api-adaptor version v0.22.0-dev
cloud-api-adaptor: starting Cloud API Adaptor daemon for "libvirt"
2026/08/27 13:43:59 [adaptor/cloud/libvirt] Created libvirt connection
2026/08/27 13:43:59 [adaptor] server config: &cloud.ServerConfig{..., PeerPodsLimitPerNode:1, ..., DeveloperMode:true}
2026/08/27 13:43:59 [util/k8sops] Successfully set extended resource for node peer-pods-worker-0
2026/08/27 13:43:59 [adaptor] server started
```
      - Worker node extended resource capacity was automatically capped at 1:
          `[kata.peerpods.io/vm](http://kata.peerpods.io/vm): 1`
      - Created `test2-devmode-pod`  , VM `podvm-test2-devmode-pod-574ddae8` was created and running on libvirt.

**Test 3: Developer Mode + Custom Limit (PEERPODS_LIMIT_PER_NODE: "2")**
    - Configuration:PEERPODS_DEVELOPER_MODE: "true"ANDPEERPODS_LIMIT_PER_NODE: "2"inpeer-pods-cm.
     Verification:
      - CAA server config confirmed:
`  server config: &cloud.ServerConfig{..., PeerPodsLimitPerNode:2, ..., DeveloperMode:true}`
      - Worker node extended resource capacity updated to 2:
`[kata.peerpods.io/vm](http://kata.peerpods.io/vm): 2`
      - Created 2 pods simultaneously (test3-pod-1 and test3-pod-2):
      - Both scheduled in parallel on the worker node.
       Libvirt ran both VMs concurrently:
```
Id   Name                         State
--------------------------------------------
1    peer-pods-ctlplane-0         running
2    peer-pods-worker-0           running
5    podvm-test3-pod-1-b18ed9eb   running
6    podvm-test3-pod-2-13ad3be1   running
```
       - Node allocated resources showed 2 / 2:
`[kata.peerpods.io/vm:2](http://kata.peerpods.io/vm): 2 2`

stevenhorsman commented:
Re-created in https://github.com/confidential-containers/cloud-api-adaptor/pull/3261 due to re-opening not being allowed
